src/Language/Haskell/TH/Lift/Generics/Internal/OuterTypeable.hs

```diff
@@ -0,0 +1,71 @@
+{-# language CPP #-}
+{-# language FlexibleInstances #-}
+{-# language UndecidableInstances #-}
+{-# language ScopedTypeVariables #-}
+#if __GLASGOW_HASKELL__ >= 708
+{-# language PolyKinds #-}
+#endif
+
+#if __GLASGOW_HASKELL__ < 710
+{-# language OverlappingInstances #-}
+#endif
+
+module Language.Haskell.TH.Lift.Generics.Internal.OuterTypeable
+  ( OuterTypeable (..)
+  ) where
+import Data.Typeable
+
+-- | A type is @OuterTypable@ if its /type constructor/ (applied to any kind
+-- arguments) is 'Typeable'. For example, @'Maybe' a@ is only 'Typeable' if @a@
+-- is 'Typeable', but it is always @OuterTypeable@.
+class OuterTypeable a where
+  -- | Get the 'TypeRep' corresponding to the outermost constructor
+  -- of a type.
+  getConTR :: proxy a -> TypeRep
+
+#if __GLASGOW_HASKELL__ >= 708
+
+# if __GLASGOW_HASKELL__ >= 710
+instance {-# OVERLAPPING #-} OuterTypeable f => OuterTypeable (f a) where
```

Comment:
Style nitpick: I'd prefer to confine the CPP to only the part that's different—namely, the `{-# OVERLAPPING #-}` bit. Something like:

```hs
instance
# if __GLASGOW_HASKELL__ >= 710
  {-# OVERLAPPING #-}
# endif
  OuterTypeable f => OuterTypeable (f a) where
    ...
```

Reply:
I tried that. I found it so difficult to read that I even prefer what I have now.